### PR TITLE
fix: better support decimal overflow

### PR DESCRIPTION
**Pull request by justindomingue**
ethersjs tryParseUnit returns '0' when parsed input has more decimal than `currency.decimal`. Especially problematic in the add liquidity flow where we snap the user input to the nearest tick, sometimes adding decimals.

### CODE REVIEW

File: src/state/swap/hooks.test.ts [hunk 1]
```diff
@@ -1,8 +1,33 @@
...
+      expect(tryParseAmount(undefined, undefined)).toBeUndefined()
+      expect(tryParseAmount('', undefined)).toBeUndefined()
+      expect(tryParseAmount('abc', undefined)).toBeUndefined()
+      expect(tryParseAmount('0', undefined)).toBeUndefined()
+    })
+
+    it('should return a CurrencyAmount', () => {
+      const currency = new Token(1, '0x6b175474e89094c44da98b954eedeac495271d0f', 6)
+
+      expect(tryParseAmount('20.05', currency)?.toSignificant()).toEqual(
```
Comment (moodysalem) [inline src/state/swap/hooks.test.ts:19]: how many sig figs is toSignificant?

**Comment by moodysalem** [comment]
can we link a bug that describes what this fixes?

File: src/state/swap/hooks.test.ts (hunk 1 above)
Comment (justindomingue) [reply src/state/swap/hooks.test.ts]: explicitly defined it

File: src/state/mint/v3/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { Token } from '@uniswap/sdk-core'
+import { tryParseAmountToPrice } from './utils'
+
+describe('hooks', () => {
+  describe('#tryParseTick', () => {
```
Comment (moodysalem) [inline src/state/mint/v3/utils.test.ts:5]: typo

File: src/state/mint/v3/utils.ts [hunk 3]
```diff
@@ -7,9 +7,23 @@ import {
...
+    return undefined
+  }
+
+  const numDecimals = value.indexOf('.') > -1 ? value.length - value.indexOf('.') - 1 : 0
+
+  return new Price(
+    baseToken,
+    quoteToken,
+    10 ** numDecimals * 10 ** baseToken.decimals,
+    parseFloat(value) * 10 ** numDecimals * 10 ** quoteToken.decimals
```
Comment (moodysalem) [inline src/state/mint/v3/utils.ts:23]: is there a better way than indexOf + parseFloat? like can you get the number of digits after the decimal from the number instance?

Comment (justindomingue) [reply src/state/mint/v3/utils.ts:23]: I think that `'.split('.')` is the simplest way to get the # of decimals.

File: src/state/mint/v3/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      expect(tryParsePrice(baseToken, quoteToken, '')).toBeUndefined()
+      expect(tryParsePrice(baseToken, quoteToken, 'abc.123')).toBeUndefined()
+      expect(tryParsePrice(baseToken, quoteToken, '1.2.3')).toBeUndefined()
+    })
+
+    it('should return a price', () => {
+      const baseToken = new Token(1, '0x6b175474e89094c44da98b954eedeac495271d0f', 6)
+      const quoteToken = new Token(1, '0x1b175474e89094c44da98b954eedeac495271d0f', 6)
+
+      expect(tryParsePrice(baseToken, quoteToken, '20')?.toSignificant(6)).toEqual('20')
```
Comment (moodysalem) [inline src/state/mint/v3/utils.test.ts:21]: `20.`?

Comment (moodysalem) [reply src/state/mint/v3/utils.test.ts:21]: extremely large/small numbers? e.g. 2^128 and 2^-128

File: src/state/mint/v3/utils.ts [hunk 5]
```diff
@@ -7,9 +7,35 @@ import {
...
-import { tryParseAmount } from 'state/swap/hooks'
 import JSBI from 'jsbi'
 
+export function tryParsePrice(baseToken?: Token, quoteToken?: Token, value?: string) {
+  if (!baseToken || !quoteToken || !value) {
+    return undefined
+  }
+
+  if (!value.match(/^[0-9,.]+$/)) {
+    console.log('regex')
```
Comment (moodysalem) [inline src/state/mint/v3/utils.ts:18]: console

File: src/state/mint/v3/utils.ts [hunk 6]
```diff
@@ -7,9 +7,35 @@ import {
...
 import { Price, Token } from '@uniswap/sdk-core'
-import { tryParseAmount } from 'state/swap/hooks'
 import JSBI from 'jsbi'
 
+export function tryParsePrice(baseToken?: Token, quoteToken?: Token, value?: string) {
+  if (!baseToken || !quoteToken || !value) {
+    return undefined
+  }
+
+  if (!value.match(/^[0-9,.]+$/)) {
```
Comment (moodysalem) [inline src/state/mint/v3/utils.ts:17]: probably more like `^\d*\.?\d+$`

File: src/state/mint/v3/utils.ts [hunk 7]
```diff
@@ -7,9 +7,35 @@ import {
...
+    console.log('regex')
+    return undefined
+  }
+
+  const [whole, fraction, ...rest] = value.split('.')
+
+  if (rest.length > 0) {
+    return undefined
```
Comment (moodysalem) [inline src/state/mint/v3/utils.ts:25]: never true with new regex

**Review by moodysalem — APPROVED** [review]

**Pull request merged**